src/mumble/GlobalShortcut_win.cpp

```diff
@@ -130,576 +142,395 @@ GlobalShortcutWin::~GlobalShortcutWin() {
+			m_xinput.reset();
 		}
 	}
-
-	// Wait for MainWindow's constructor to finish before we enumerate DirectInput devices.
-	// We need to do this because adding a new device requires a Window handle. (SetCooperativeLevel())
-	while (!Global::get().mw)
-		this->yieldCurrentThread();
-
+#endif
 #ifdef USE_GKEY
 	if (Global::get().s.bEnableGKey) {
-		gkey = new GKeyLibrary();
-		qWarning("GlobalShortcutWin: GKeys initialized, isValid: %d", gkey->isValid());
+		m_gkey.reset(new GKeyLibrary());
+		if (!m_gkey->isValid()) {
+			qWarning("GlobalShortcutWin: Failed to initialize GKey!");
+			m_gkey.reset();
+		}
 	}
 #endif
-
-#ifdef USE_XBOXINPUT
-	if (Global::get().s.bEnableXboxInput) {
-		xboxinput = new XboxInput();
-		ZeroMemory(&xboxinputLastPacket, sizeof(xboxinputLastPacket));
-		qWarning("GlobalShortcutWin: XboxInput initialized, isValid: %d", xboxinput->isValid());
+	// Wait for MainWindow's constructor to finish, so that we can get the window's handle.
+	while (!mumble_mw_hwnd) {
+		yieldCurrentThread();
 	}
-#endif
 
-	QTimer *timer = new QTimer;
-	connect(timer, SIGNAL(timeout()), this, SLOT(timeTicked()));
-	timer->start(20);
+	RAWINPUTDEVICE rid[5];
 
-	setPriority(QThread::TimeCriticalPriority);
+	rid[0].usUsagePage = HID_USAGE_PAGE_GENERIC;
+	rid[0].usUsage     = HID_USAGE_GENERIC_MOUSE;
+	rid[0].dwFlags     = RIDEV_INPUTSINK;
+	rid[0].hwndTarget  = mumble_mw_hwnd;
 
-	exec();
+	rid[1].usUsagePage = HID_USAGE_PAGE_GENERIC;
+	rid[1].usUsage     = HID_USAGE_GENERIC_JOYSTICK;
+	rid[1].dwFlags     = RIDEV_INPUTSINK | RIDEV_DEVNOTIFY;
+	rid[1].hwndTarget  = mumble_mw_hwnd;
 
-	delete timer;
+	rid[2].usUsagePage = HID_USAGE_PAGE_GENERIC;
+	rid[2].usUsage     = HID_USAGE_GENERIC_GAMEPAD;
+	rid[2].dwFlags     = RIDEV_INPUTSINK | RIDEV_DEVNOTIFY;
+	rid[2].hwndTarget  = mumble_mw_hwnd;
 
-#ifdef USE_GKEY
-	delete gkey;
-#endif
+	rid[3].usUsagePage = HID_USAGE_PAGE_GENERIC;
+	rid[3].usUsage     = HID_USAGE_GENERIC_KEYBOARD;
+	rid[3].dwFlags     = RIDEV_INPUTSINK | RIDEV_DEVNOTIFY;
+	rid[3].hwndTarget  = mumble_mw_hwnd;
 
-#ifdef USE_XBOXINPUT
-	delete xboxinput;
-#endif
+	rid[4].usUsagePage = HID_USAGE_PAGE_GENERIC;
+	rid[4].usUsage     = HID_USAGE_GENERIC_MULTI_AXIS_CONTROLLER;
+	rid[4].dwFlags     = RIDEV_INPUTSINK | RIDEV_DEVNOTIFY;
+	rid[4].hwndTarget  = mumble_mw_hwnd;
 
-	if (bHook) {
-		if (hhMouse) {
-			UnhookWindowsHookEx(hhMouse);
-		}
-		if (hhKeyboard) {
-			UnhookWindowsHookEx(hhKeyboard);
-		}
+	if (!RegisterRawInputDevices(rid, sizeof(rid) / sizeof(rid[0]), sizeof(rid[0]))) {
+		qWarning("GlobalShortcutWindows: RegisterRawInputDevices() failed with error %u", GetLastError());
 	}
 
-	foreach (InputDevice *id, qhInputDevices) {
-		if (id->pDID) {
-			id->pDID->Unacquire();
-			id->pDID->Release();
-		}
-		delete id;
-	}
-	pDI->Release();
+	QTimer timer;
+	connect(&timer, &QTimer::timeout, this, &GlobalShortcutWin::timeTicked);
+	timer.start(20);
+
+	setPriority(QThread::TimeCriticalPriority);
+
+	exec();
```

Comment:
Does `exec` block long enough for the timer to complete? And does it matter if the scope is exited while the timer is running?

Reply:
From https://doc.qt.io/qt-5/qthread.html#exec:

> Enters the event loop and waits until [exit](https://doc.qt.io/qt-5/qthread.html#exit)() is called, returning the value that was passed to [exit](https://doc.qt.io/qt-5/qthread.html#exit)(). The value returned is 0 if [exit](https://doc.qt.io/qt-5/qthread.html#exit)() is called via [quit](https://doc.qt.io/qt-5/qthread.html#quit)().
>
> This function is meant to be called from within [run](https://doc.qt.io/qt-5/qthread.html#run)(). It is necessary to call this function to start event handling.
>
> Note: This can only be called within the thread itself, i.e. when it is the current thread.

Right now we're injecting input messages we receive from Windows instead of using a proper signal because we need the return value (`false`) or (`true`) in order to know whether we should prevent the message from being processed further:

https://github.com/mumble-voip/mumble/blob/f8303557b729eca060b536aa6cbd48530fea8583/src/mumble/MumbleApplication.cpp#L59-L86

https://github.com/mumble-voip/mumble/blob/f8303557b729eca060b536aa6cbd48530fea8583/src/mumble/MumbleApplication.cpp#L88-L97

From https://doc.qt.io/qt-5/qabstractnativeeventfilter.html#nativeEventFilter:

> In your reimplementation of this function, if you want to filter the message out, i.e. stop it being handled further, return true; otherwise return false.

`WM_INPUT` messages are not processed as input, in fact specifying `RIDEV_NOLEGACY` for a `RAWINPUTDEVICE` structure passed to [`RegisterRawInputDevices()`](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-registerrawinputdevices) results in keystrokes and mouse clicks not being handled at all.

In conclusion: we can get rid of the event system and use signal + slot, so that we don't block the window thread.

As for the `QTimer`, from https://doc.qt.io/qt-5/qtimer.html#details:

> In multithreaded applications, you can use QTimer in any thread that has an event loop. To start an event loop from a non-GUI thread, use [QThread::exec](https://doc.qt.io/qt-5/qthread.html#exec)(). Qt uses the timer's [thread affinity](https://doc.qt.io/qt-5/qobject.html#thread) to determine which thread will emit the [timeout](https://doc.qt.io/qt-5/qtimer.html#timeout)() signal. Because of this, you must start and stop the timer in its thread; it is not possible to start a timer from another thread.